# Statistical Methods in Pandas - Lab

## Introduction

In this lesson you'll get some hands on experience using some of the key summary statistics methods in Pandas.

## Objectives:
You will be able to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets.  You will find this dataset in the file `lego_sets.csv`.  

In the cell below:

* Import pandas and set the standard alias of `pd`
* Load in the `lego_sets.csv`dataset using the `read_csv()` function
* Display the head of the DataFrame to get a feel for what we'll be working with

In [1]:
import pandas as pd1

In [6]:
df = pd.read_csv('lego_sets.csv')
df.head(100)

ages  list_price  num_reviews  piece_count  play_star_rating  \
0   6-12       29.99          2.0        277.0               4.0   
1   6-12       19.99          2.0        168.0               4.0   
2   6-12       12.99         11.0         74.0               4.3   
3    12+       99.99         23.0       1032.0               3.6   
4    12+       79.99         14.0        744.0               3.2   
5    12+       59.99          7.0        597.0               3.7   
6    12+       59.99         37.0        598.0               3.7   
7    12+       49.99         24.0        780.0               4.4   
8    12+       39.99         23.0        468.0               3.6   
9    12+       39.99         11.0        444.0               3.6   
10   12+       39.99         14.0        386.0               4.1   
11   12+       34.99         53.0        321.0               3.2   
12   12+       29.99          7.0        361.0               4.2   
13  7-12      159.99         63.0        847.0               3.8   
14   10+       29.99         13.0        708.0               4.7   
15   10+       19.99          1.0        234.0               3.0   
16   10+       19.99          1.0        160.0               5.0   
17   10+        9.99          1.0        149.0               2.0   
18   10+        9.99          1.0        141.0               2.0   
19   10+        9.99          3.0        101.0               4.0   
20   10+        9.99          2.0        105.0               3.0   
21   10+        9.99          1.0        113.0               5.0   
22   10+        9.99          1.0        136.0               NaN   
23   10+        9.99          2.0         91.0               3.0   
24   10+        9.99          7.0        140.0               3.2   
25   10+        9.99          5.0        143.0               4.6   
26   10+        9.99          3.0        122.0               2.7   
27   10+        9.99          5.0        130.0               4.3   
28   10+        9.99          3.0        119.0               4.5   
29   10+        9.99          1.0        135.0               1.0   
..   ...         ...          ...          ...               ...   
70  6-12       39.99          5.0        378.0               4.8   
71  5-12       39.99          8.0        257.0               4.2   
72  5-12       29.99          7.0        344.0               5.0   
73  5-12       29.99          7.0        310.0               4.4   
74  5-12       24.99         11.0        214.0               4.5   
75  6-12       19.99          4.0        192.0               5.0   
76  5-12       19.99         35.0        249.0               4.6   
77  5-12       19.99         89.0         24.0               2.9   
78  5-12       19.99         12.0        239.0               4.7   
79  5-12       19.99          6.0        190.0               4.2   
80  5-12       19.99          7.0        174.0               4.3   
81  5-12       19.99          8.0        144.0               4.4   
82  5-12       19.99          1.0        126.0               5.0   
83  5-12       19.99          1.0        127.0               5.0   
84   4-7       19.99          NaN        115.0               NaN   
85  5-12       19.99          3.0        141.0               4.3   
86  6-12       19.99          5.0        201.0               4.3   
87  5-12       15.99         47.0          8.0               4.2   
88  5-12       14.99         85.0          2.0               4.1   
89  5-12       14.99         40.0          2.0               4.1   
90  5-12        9.99          2.0         87.0               4.5   
91  5-12        9.99         12.0         88.0               4.5   
92   4-7        9.99          NaN         73.0               NaN   
93  5-12        9.99          4.0         76.0               4.8   
94  5-12        9.99          2.0         88.0               NaN   
95  5-12        9.99          2.0         82.0               4.5   
96  5-12        9.99          1.0         78.0              

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset.  There are two ways we'll get this information-- `.info()` and `.describe()`.

### Using `.info()`

The `.info()` method provides us metadata on the DataFrame itself.  This allows to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset? 
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null float64
num_reviews          10641 non-null float64
piece_count          12261 non-null float64
play_star_rating     10486 non-null float64
prod_desc            11884 non-null object
prod_id              12261 non-null float64
prod_long_desc       12261 non-null object
review_difficulty    10206 non-null object
set_name             12261 non-null object
star_rating          10641 non-null float64
theme_name           12258 non-null object
val_star_rating      10466 non-null float64
country              12261 non-null object
dtypes: float64(7), object(7)
memory usage: 1.3+ MB


#### Interpreting the Results

Read the output above, and then answer the following questions:

How many total rows are in this DataFrame?  How many columns contain numeric data? How many contain categorical data?  Identify at least 3 columns that contain missing values. 

Write your answer below this line:

_______________________________________________________________________________________________________________________________

12261 rows, 7  columns contain numerical data, 7 contain categorical data, missing values: player_star_rating, review_difficulty, and star_rating

There are 12261 rows, as evidenced by the Range Index listed in the output.  Of the 14 columns in this dataset, 7 contain numeric data stored as `float64`, and the other 7 contain strings, which pandas lists as `object`.  The column `num_reviews`, `review_difficulty`, and `val_star_rating` all contain missing values.  We can tell this because the number of entries for each of those columns is less than the total number of entries described by the Range Index. 



## Using `.describe()`

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method. 

In [9]:
df.describe()
#df['val_star_rating'].median()

list_price   num_reviews   piece_count  play_star_rating  \
count  12261.000000  10641.000000  12261.000000      10486.000000   
mean      65.141998     16.826238    493.405921          4.337641   
std       91.980429     36.368984    825.364580          0.652051   
min        2.272400      1.000000      1.000000          1.000000   
25%       19.990000      2.000000     97.000000          4.000000   
50%       36.587800      6.000000    216.000000          4.500000   
75%       70.192200     13.000000    544.000000          4.800000   
max     1104.870000    367.000000   7541.000000          5.000000   

            prod_id   star_rating  val_star_rating  
count  1.226100e+04  10641.000000     10466.000000  
mean   5.983675e+04      4.514134         4.228960  
std    1.638115e+05      0.518865         0.660282  
min    6.300000e+02      1.800000         1.000000  
25%    2.103400e+04      4.300000         4.000000  
50%    4.206900e+04      4.700000         4.300000  
75%    7.092200e+04      5.000000         4.700000  
max    2.000431e+06      5.000000         5.000000

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns.  Use these to answer the following questions:

How much is the standard deviation for piece count?  How many pieces are in the largest lego set?  How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

1. 825.364580

2. 7541

3. 1

4. 4.3

The standard deviation is 825 pieces.  The largest lego set contains 7,541 pieces, while the smallest contains only 1.  The median value for `val_star_rating` is 4.3.

## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods. 

In the cell below, compute the median value of the `star_rating` column.

In [10]:
df['star_rating'].median()

4.7

Next, get a count of the total number of values in `play_star_rating`.

In [15]:
len(df['play_star_rating'].unique())

31

Now, compute the standard deviation of the `list_price` column.

In [16]:
df['list_price'].std()

91.9804293059252

If we bought every single lego set in this dataset, how many pieces would we have?  Use the `.sum()` method on the correct column to compute this. 

In [18]:
df['piece_count'].sum()

6049650.0

Now, let's try getting the value for the 90% quantile.  Do this in the cell below.

In [19]:
df['piece_count'].quantile(.9)

1077.0

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data--there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data. 

In the cell below, get the `.unique()` values contained within the `review_difficulty` column. 

In [20]:
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts` for this column, to see how common each is. 

In [21]:
df['review_difficulty'].value_counts()

Easy                4236
Average             3765
Very Easy           1139
Challenging         1058
Very Challenging       8
Name: review_difficulty, dtype: int64

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `applymap()` function and passing in a lambda function. 

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call our DataFrame's `.applymap()` function and pass in `lambda x: str(x)`
* Call our new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [23]:
string_df = df.applymap(lambda x: str(x))
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12261 entries, 0 to 12260
Data columns (total 14 columns):
ages                 12261 non-null object
list_price           12261 non-null object
num_reviews          12261 non-null object
piece_count          12261 non-null object
play_star_rating     12261 non-null object
prod_desc            12261 non-null object
prod_id              12261 non-null object
prod_long_desc       12261 non-null object
review_difficulty    12261 non-null object
set_name             12261 non-null object
star_rating          12261 non-null object
theme_name           12261 non-null object
val_star_rating      12261 non-null object
country              12261 non-null object
dtypes: object(14)
memory usage: 1.3+ MB


Note that everything--even the `NaN` values, have been cast to a string in the example above. 

Note that for pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame--this will become especially useful when we learn how to **_normalize_** our datasets in a later section!

# Conclusion

In this lab, we learned how to:

* Understand and use the df.describe() and df.info() summary statistics methods
* Use built-in Pandas methods for calculating summary statistics (.mean(), .std(), .count(), .sum(), .mean(), .median(), .std(), .var() and .quantile())
* Apply a function to every element in a Series or DataFrame using s.apply() and df.applymap()